pandas/core/indexes/base.py

```diff
@@ -2728,41 +2728,35 @@ def _union(self, other, sort):
         lvals = self._values
         rvals = other._values
 
-        if sort is None and self.is_monotonic and other.is_monotonic:
-            try:
-                result = self._outer_indexer(lvals, rvals)[0]
-            except TypeError:
-                # incomparable objects
-                result = list(lvals)
-
-                # worth making this faster? a very unusual case
-                value_set = set(lvals)
-                result.extend([x for x in rvals if x not in value_set])
-                result = Index(result)._values  # do type inference here
+        try:
+            # If one of both is not monotonic, we resort it afterwards
+            result = self._outer_indexer(np.sort(lvals), np.sort(rvals))[0]
+        except TypeError:
+            # incomparable objects
+            result = list(lvals)
+
+            # worth making this faster? a very unusual case
+            value_set = set(lvals)
+            result.extend([x for x in rvals if x not in value_set])
+            result = Index(result)._values  # do type inference here
         else:
-            # find indexes of things in "other" that are not in "self"
-            if self.is_unique:
-                indexer = self.get_indexer(other)
-                indexer = (indexer == -1).nonzero()[0]
-            else:
-                indexer = algos.unique1d(self.get_indexer_non_unique(other)[1])
-
-            if len(indexer) > 0:
-                other_diff = algos.take_nd(rvals, indexer, allow_fill=False)
-                result = concat_compat((lvals, other_diff))
-
-            else:
-                result = lvals
-
-            if sort is None:
-                try:
-                    result = algos.safe_sort(result)
-                except TypeError as err:
-                    warnings.warn(
-                        f"{err}, sort order is undefined for incomparable objects",
-                        RuntimeWarning,
-                        stacklevel=3,
-                    )
+            if (
+                sort is False
+                or not self.is_monotonic
+                or not other.is_monotonic
+                or is_categorical_dtype(self)
+            ):
+                result = algos.re_sort_union_after_inputs(result, lvals, rvals)
+
+        if sort is None and (not self.is_monotonic or not other.is_monotonic):
+            try:
+                result = algos.safe_sort(result)
+            except TypeError as err:
+                warnings.warn(
+                    f"{err}, sort order is undefined for incomparable objects",
+                    RuntimeWarning,
+                    stacklevel=3,
+                )
```

Comment:
potential separate PR: i think there are 4 places in this file where we do this safe_sort call, and we are not consistent in a) issuing this warning or b) catching the TypeError.  Could be refactored into a small helper method.

Reply:
Will look into this in the coming days